drivers/gpio/gpio_rcar.c

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright (c) 2020 IoT.bzh
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT renesas_rcar_gpio
+
+#include <errno.h>
+
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <drivers/gpio.h>
+#include <drivers/clock_control.h>
+#include <drivers/clock_control/rcar_clock_control.h>
+
+#include "gpio_utils.h"
+
+typedef void (*init_func_t)(const struct device *dev);
+
+struct gpio_rcar_cfg {
+	struct gpio_driver_config common;
+	uint32_t reg_addr;
+	int reg_size;
+	init_func_t init_func;
+	char *clock_controller;
+	struct rcar_cpg_clk mod_clk;
+};
+
+struct gpio_rcar_data {
+	struct gpio_driver_data common;
+	sys_slist_t cb;
+};
+
+#define IOINTSEL 0x00   /* General IO/Interrupt Switching Register */
+#define INOUTSEL 0x04   /* General Input/Output Switching Register */
+#define OUTDT    0x08   /* General Output Register */
+#define INDT     0x0c   /* General Input Register */
+#define INTDT    0x10   /* Interrupt Display Register */
+#define INTCLR   0x14   /* Interrupt Clear Register */
+#define INTMSK   0x18   /* Interrupt Mask Register */
+#define MSKCLR   0x1c   /* Interrupt Mask Clear Register */
+#define POSNEG   0x20   /* Positive/Negative Logic Select Register */
+#define EDGLEVEL 0x24   /* Edge/level Select Register */
+#define FILONOFF 0x28   /* Chattering Prevention On/Off Register */
+#define OUTDTSEL 0x40   /* Output Data Select Register */
+#define BOTHEDGE 0x4c   /* One Edge/Both Edge Select Register */
```

Comment:
These values should ideally be provided by a vendor header file from a board support package. We keep these HAL support files in `../modules/hal/`. The name of the driver is `gpio_rcar` and that implies it should support all Renesans "R-Car" family SoCs. I'm not familiar with "R-Car" however typically register addresses change between SoC series. The one defined in this driver are tailored for R-Car-H3 SoC.

Reply:
On RCar SoC these registers are almost compatible for R-Car GEN 1,2,3.
From what I see in Linux If one want to add support for Gen1 and Gen2, only the two last registers 
OUTDTSEL and BOTHEDGE may not be available.

Reply:
As per #33171 Renesas doesn't provide its own HAL, we can keep register defines directly in the driver.